File: getting_started/dotnet/getting_started/MQTTnet.Client.Extensions/ProtobufSerializer.cs

using Google.Protobuf;

namespace MQTTnet.Client.Extensions;

public class ProtobufSerializer : IMessageSerializer
{
    private readonly MessageParser? _parser;

    public ProtobufSerializer(MessageParser parser) => _parser = parser;

    public string ContentType => "application/protobuf";

    public T FromBytes<T>(byte[] payload) => (T)_parser!.ParseFrom(payload);

    public byte[] ToBytes<T>(T payload) => (payload as IMessage).ToByteArray();

}

File: scenarios/command/ts/protoMessages/src/protobufSerializer.ts
import {
    IMessageSerializer
} from '@mqttapplicationsamples/mqttjsclientextensions';

export class ProtobufSerializer implements IMessageSerializer {
    private protoObject: any;

    constructor(protoObject: any) {
        this.protoObject = protoObject;
    }

    public contentType = "application/protobuf";

    public fromBytes<T>(payload: Buffer): T {
        return this.protoObject.decode(payload) as T;
    }

    public toBytes<T>(payload: T): Buffer {
        return Buffer.from(this.protoObject.encode(payload).finish());
    }
}
